### PR TITLE
Align immigration status with SSN card type

### LINKED ISSUE #337: Align immigration status with SSN card type

### PR DESCRIPTION
**Pull request by daphnehanse11**
Fixes #337

### CODE REVIEW

File: policyengine_us_data/datasets/cps/cps.py [hunk 1]
```diff
@@ -1345,6 +1345,42 @@ def select_random_subset_to_target(
...
+    imm_class[mask] = b"DACA"
+
+    mask = (ssn_card_type == 3) & (years_in_us <= 5) & (imm_class == b"UNSET")
+    imm_class[mask] = b"HUMANITARIAN_RECENT"
+
+    mask = (ssn_card_type == 2) & (imm_class == b"UNSET")
+    imm_class[mask] = b"TEMP_NONQUALIFIED"
+
+    mask = (ssn_card_type == 3) & (imm_class == b"UNSET")
+    imm_class[mask] = b"LPR_OR_QUALIFIED"
```
Comment (PavelMakarchuk) [inline policyengine_us_data/datasets/cps/cps.py:1381]: Do we need to make an adjustment to `-us` here? We dont have some of these in the [enum](https://github.com/PolicyEngine/policyengine-us/blob/master/policyengine_us/variables/household/demographic/person/immigration_status.py)

Comment (daphnehanse11) [reply policyengine_us_data/datasets/cps/cps.py]: COFA matters much less since there is no policy change, but to bundle DACA and TPS as one status is inaccurate as they have different healthcare eligibility rules.

Comment (MaxGhenis) [reply policyengine_us_data/datasets/cps/cps.py]: ok can you make the change in -us

Comment (PavelMakarchuk) [reply policyengine_us_data/datasets/cps/cps.py]: This will affect multiple programs and benefit navigator household API calls so need to communicate with them before merging an `-us` update

Comment (MaxGhenis) [reply policyengine_us_data/datasets/cps/cps.py]: we can temporarily keep DACA/TPS as a category with the current logic while also adding them separately

File: policyengine_us_data/datasets/cps/cps.py [hunk 2]
```diff
@@ -1345,6 +1345,64 @@ def select_random_subset_to_target(
             "population": code_0_after,
         }
     )
+    # NEW IMMIGRATION-STATUS TAGS FOR OBFBA
+    # ------------------------------------
+    years_in_us = 2024 - (1981 + person.PEINUSYR)
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1350]: make 2024 and 1981 constants (uppercase)

File: policyengine_us_data/datasets/cps/cps.py [hunk 3]
```diff
@@ -1345,6 +1345,64 @@ def select_random_subset_to_target(
...
+    # start every non-citizen as LPR so no UNSET survives
+    immigration_status = np.full(
+        len(person), "LEGAL_PERMANENT_RESIDENT", dtype="U32"
+    )
+
+    # 1. Undocumented (Code 0)
+    immigration_status[ssn_card_type == 0] = "UNDOCUMENTED"
+
+    # 2. COFA migrants – treat as LPR (kept from your logic, optional)
+    cofa_codes = {316, 317, 329}
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1363]: is any of this logic shared with ssn_card_type?

File: policyengine_us_data/tests/test_datasets/test_enhanced_cps.py [hunk 4]
```diff
@@ -105,6 +105,30 @@ def test_ssn_card_type_none_target():
     assert pct_error < TOLERANCE
 
 
+def test_immigration_status_none_target():
```
Comment (MaxGhenis) [inline policyengine_us_data/tests/test_datasets/test_enhanced_cps.py:108]: similarly can we unify the test with ssn_card_type?

Comment (daphnehanse11) [reply policyengine_us_data/tests/test_datasets/test_enhanced_cps.py:108]: How? Don't we want to ensure the new sorting doesn't disrupt the old categories?

File: policyengine_us_data/datasets/cps/cps.py [hunk 5]
```diff
@@ -1345,6 +1345,69 @@ def select_random_subset_to_target(
             "population": code_0_after,
         }
     )
+    # NEW IMMIGRATION-STATUS TAGS FOR OBFBA
+
+    CURRENT_YEAR = 2024
+    AMNESTY_YEAR = (
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1351]: comment above to avoid (

File: policyengine_us_data/datasets/cps/cps.py [hunk 6]
```diff
@@ -1345,6 +1345,69 @@ def select_random_subset_to_target(
...
+        1981  # Immigrants who entered before 1982 are eligible for amnesty
+    )
+    # ------------------------------------
+    years_in_us = CURRENT_YEAR - (1981 + person.PEINUSYR)
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1355]: use the 1981 constant

File: policyengine_us_data/datasets/cps/cps.py [hunk 7]
```diff
@@ -1345,6 +1345,69 @@ def select_random_subset_to_target(
...
+    # start every non-citizen as LPR so no UNSET survives
+    immigration_status = np.full(
+        len(person), "LEGAL_PERMANENT_RESIDENT", dtype="U32"
+    )
+
+    # 1. Undocumented (Code 0)
+    immigration_status[ssn_card_type == 0] = "UNDOCUMENTED"
+
+    # 2. COFA migrants – treat as LPR (kept from your logic, optional)
+    cofa_codes = {316, 317, 329}
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1368]: all constants
```suggestion
    COFA_CODES = {316, 317, 329}
```

File: policyengine_us_data/datasets/cps/cps.py [hunk 8]
```diff
@@ -1345,6 +1345,69 @@ def select_random_subset_to_target(
...
+    immigration_status[mask] = "LEGAL_PERMANENT_RESIDENT"
+
+    # 3. Cuban / Haitian entrants (<=10 yrs in US)
+    mask = (
+        (ssn_card_type != 0) & np.isin(birth, [241, 250]) & (years_in_us <= 10)
+    )
+    immigration_status[mask] = "CUBAN_HAITIAN_ENTRANT"
+
+    # 4. DACA (came <16, now ≥18, ≥8 yrs in US, valid EAD)
+    mask = (
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/cps/cps.py:1379]: make constants for each number here

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
(note the PEINUSYR is a code not year)

File: policyengine_us_data/datasets/cps/cps.py [hunk 9]
```diff
@@ -1357,7 +1416,7 @@ def select_random_subset_to_target(
         3: "OTHER_NON_CITIZEN",  # Non-citizens with indicators of legal status
     }
     ssn_card_type_str = (
-        pd.Series(ssn_card_type).map(code_to_str).astype("S").values
+        pd.Series(ssn_card_type).map(code_to_str).astype("U").values
```
Comment (vahid-ahmadi) [inline policyengine_us_data/datasets/cps/cps.py:1419]: Do we need to change ```.astype("S")```?

File: policyengine_us_data/datasets/cps/cps.py [hunk 10]
```diff
@@ -1345,6 +1345,65 @@ def select_random_subset_to_target(
             "population": code_0_after,
         }
     )
+    # NEW IMMIGRATION-STATUS TAGS FOR OBFBA
+    CURRENT_YEAR = 2024
+    years_in_us = CURRENT_YEAR - (1981 + person.PEINUSYR)
```
Comment (vahid-ahmadi) [inline policyengine_us_data/datasets/cps/cps.py:1350]: Validate ```years_in_us``` calculation doesn't produce negative values
```years_in_us = np.maximum(0, CURRENT_YEAR - (1981 + person.PEINUSYR))```

Also add checks that ```age_at_entry``` is reasonable
```age_at_entry = np.maximum(0, person.A_AGE - years_in_us)```

File: policyengine_us_data/tests/test_datasets/test_enhanced_cps.py [hunk 11]
```diff
@@ -105,6 +105,34 @@ def test_ssn_card_type_none_target():
     assert pct_error < TOLERANCE
 
 
+def test_undocumented_matches_ssn_none():
```
Comment (vahid-ahmadi) [inline policyengine_us_data/tests/test_datasets/test_enhanced_cps.py:108]: Can we add some tests for edge cases? Some cases like these:
- A person who entered the US at exactly age 16 should not qualify for DACA.
- A person who entered 10 years ago should still be eligible for Cuban/Haitian entrant logic.

File: policyengine_us_data/datasets/cps/cps.py [hunk 12]
```diff
@@ -1345,6 +1345,65 @@ def select_random_subset_to_target(
...
+    mask = (ssn_card_type != 0) & np.isin(birth, list(cofa_codes))
+    immigration_status[mask] = "LEGAL_PERMANENT_RESIDENT"
+
+    # 3. Cuban / Haitian entrants (≤ 10 yrs in US)
+    mask = (
+        (ssn_card_type != 0) & np.isin(birth, [241, 250]) & (years_in_us <= 10)
+    )
+    immigration_status[mask] = "CUBAN_HAITIAN_ENTRANT"
+
+    # 4. DACA (came <16, now ≥18, ≥8 yrs in US, valid EAD)
```
Comment (vahid-ahmadi) [inline policyengine_us_data/datasets/cps/cps.py:1375]: For better readability, can we move this part to a function like ```classify_daca_eligible```?
